service/src/main/java/bio/terra/pipelines/stairway/imputation/gcp/PollCromwellSubmissionStatusStep.java

```diff
@@ -0,0 +1,106 @@
+  private final RawlsService rawlsService;
+  private final SamService samService;
+  private final ImputationConfiguration imputationConfiguration;
+  private final Logger logger = LoggerFactory.getLogger(PollCromwellSubmissionStatusStep.class);
+
+  public PollCromwellSubmissionStatusStep(
+      SamService samService,
+      RawlsService rawlsService,
+      ImputationConfiguration imputationConfiguration) {
+    this.samService = samService;
+    this.rawlsService = rawlsService;
+    this.imputationConfiguration = imputationConfiguration;
+  }
+
+  @Override
+  public StepResult doStep(FlightContext flightContext) throws InterruptedException {
+    // validate and extract parameters from input map
+    FlightMap inputParameters = flightContext.getInputParameters();
+    FlightUtils.validateRequiredEntries(
+        inputParameters,
+        RunImputationJobFlightMapKeys.CONTROL_WORKSPACE_NAME,
+        RunImputationJobFlightMapKeys.CONTROL_WORKSPACE_PROJECT);
+    String controlWorkspaceName =
+        inputParameters.get(RunImputationJobFlightMapKeys.CONTROL_WORKSPACE_NAME, String.class);
+    String controlWorkspaceProject =
+        inputParameters.get(RunImputationJobFlightMapKeys.CONTROL_WORKSPACE_PROJECT, String.class);
+    // validate and extract parameters from working map
+    FlightMap workingMap = flightContext.getWorkingMap();
+    FlightUtils.validateRequiredEntries(workingMap, RunImputationJobFlightMapKeys.SUBMISSION_ID);
+
+    UUID submissionId = workingMap.get(RunImputationJobFlightMapKeys.SUBMISSION_ID, UUID.class);
+
+    // poll until all runs are in a finalized state
+    Submission submissionResponse = null;
+    boolean stillRunning = true;
+    try {
+      while (stillRunning) {
+        submissionResponse =
+            rawlsService.getSubmissionStatus(
+                samService.getTeaspoonsServiceAccountToken(),
+                controlWorkspaceProject,
+                controlWorkspaceName,
+                submissionId);
+        stillRunning = RawlsService.submissionIsRunning(submissionResponse);
+        if (stillRunning) {
+          logger.info(
+              "Polling Started, sleeping for {} seconds",
+              imputationConfiguration.getCromwellSubmissionPollingIntervalInSeconds());
+          TimeUnit.SECONDS.sleep(
+              imputationConfiguration.getCromwellSubmissionPollingIntervalInSeconds());
+        }
+      }
+    } catch (RawlsServiceApiException e) {
+      return new StepResult(StepStatus.STEP_RESULT_FAILURE_RETRY, e);
+    }
+
+    // if there are any non-successful workflows, fatally fail the step
+    List<Workflow> failedRunLogs =
+        submissionResponse.getWorkflows().stream()
+            .filter(workflow -> !workflow.getStatus().equals(WorkflowStatus.SUCCEEDED))
+            .toList();
+    if (failedRunLogs.isEmpty()) {
+      return StepResult.getStepResultSuccess();
+    } else {
+      return new StepResult(
+          StepStatus.STEP_RESULT_FAILURE_FATAL,
```

Comment:
if we do a fatal failure here, do we need to increment the failed run counter here too? alternatively we can do a non-fatal failure so we run through the undo steps (i don't think we actually undo anything) so we get to the failed run counter increment that we will add to the undoStep method of whatever ends up being the first step.